### DISCUSSION TITLE
Why do you compile k3s again in package-cli when make  command

### DISCUSSION THREAD

**Discussion by sunnoy** [k3s-io/k3s#10491] [Q&A] [answered]
- first build

scripts/build

```bash
echo Building k3s
CGO_ENABLED=1 "${GO}" build $BLDFLAGS -tags "$TAGS" -buildvcs=false -gcflags="all=${GCFLAGS}" -ldflags "$VERSIONFLAGS $LDFLAGS $STATIC" -o bin/k3s${BINARY_POSTFIX} ./cmd/server

```

- second build

scripts/package-cli
```bash
CMD_NAME=dist/artifacts/k3s${BIN_SUFFIX}${BINARY_POSTFIX}

GOOS=linux CC=gcc CXX=g++ "${GO}" generate
LDFLAGS="
    -X github.com/k3s-io/k3s/pkg/version.Version=$VERSION
    -X github.com/k3s-io/k3s/pkg/version.GitCommit=${COMMIT:0:8}
    -w -s
"
TAGS="urfave_cli_no_docs"
STATIC="-extldflags '-static'"
CGO_ENABLED=0 "${GO}" build -tags "$TAGS" -buildvcs=false -ldflags "$LDFLAGS $STATIC" -o ${CMD_NAME} ./cmd/k3s
```

**Comment by brandond** [comment, marked as answer]
There are two k3s binaries:
* the self-extracting wrapper that we ship as our GH artifact
* the actual k3s binary that we ship in our docker image, and as the payload for the self-extracting wrapper

If you look at `/var/lib/rancher/k3s/data/HASH/bin` on a k3s node, you'll find another k3s binary there - this is the actual entrypoint for the running k3s process. The one we install to /usr/local/bin/k3s just extracts and calls the actual binaries.

**Reply by sunnoy** [reply]
so The k3s compiled by scripts/package-cli will be placed in /usr/local/bin/k3s, and the k3s compiled by scripts/build will be placed in /var/lib/rancher/k3s/data/HASH/bin
Can you tell us the reason for this design?

**Reply by brandond** [reply]
because we wanted to ship a single binary that contains a bunch of other binaries.
